src/highdicom/seg/sop.py

```diff
@@ -1328,35 +1387,51 @@ def from_dataset(
 
-        Builds a lookup table mapping the SOPInstanceUIDs of all datasets
-        referenced in the segmentation to a tuple containing the
-        StudyInstanceUID, SeriesInstanceUID and SOPInstanceUID.
+        Builds a table in the sqlite db containing all Study, Series and SOP
+        Instance UIDS for instances referenced in the segmentation.
 
         """
-        # Map sop uid to tuple (study uid, series uid, sop uid)
-        self._ref_ins_lut = OrderedDict()
+        cur = self._db.cursor()
+        cur.execute(
+            """
+                CREATE TABLE InstanceUIDs(
+                    SOPInstanceUID_id INTEGER PRIMARY KEY AUTOINCREMENT,
+                    StudyInstanceUID VARCHAR NOT NULL,
+                    SeriesInstanceUID VARCHAR NOT NULL,
+                    SOPInstanceUID VARCHAR UNIQUE NOT NULL
+                )
+            """
+        )
+        instance_data = []
         if hasattr(self, 'ReferencedSeriesSequence'):
             for ref_series in self.ReferencedSeriesSequence:
                 for ref_ins in ref_series.ReferencedInstanceSequence:
-                    self._ref_ins_lut[ref_ins.ReferencedSOPInstanceUID] = (
-                        hd_UID(self.StudyInstanceUID),
-                        hd_UID(ref_series.SeriesInstanceUID),
-                        hd_UID(ref_ins.ReferencedSOPInstanceUID)
+                    instance_data.append(
+                        (
+                            self.StudyInstanceUID,
+                            ref_series.SeriesInstanceUID,
+                            ref_ins.ReferencedSOPInstanceUID
+                        )
                     )
         other_studies_kw = 'StudiesContainingOtherReferencedInstancesSequence'
         if hasattr(self, other_studies_kw):
             for ref_study in getattr(self, other_studies_kw):
                 for ref_series in ref_study.ReferencedSeriesSequence:
                     for ref_ins in ref_series.ReferencedInstanceSequence:
-                        self._ref_ins_lut[ref_ins.ReferencedSOPInstanceUID] = (
-                            hd_UID(ref_study.StudyInstanceUID),
-                            hd_UID(ref_series.SeriesInstanceUID),
-                            hd_UID(ref_ins.ReferencedSOPInstanceUID)
+                        instance_data.append(
+                            (
+                                ref_study.StudyInstanceUID,
+                                ref_series.SeriesInstanceUID,
+                                ref_ins.ReferencedSOPInstanceUID,
+                            )
                         )
-
-        self._source_sop_instance_uids = np.array(
-            list(self._ref_ins_lut.keys())
+        cur.executemany(
+            "INSERT INTO InstanceUIDs "
+            "(StudyInstanceUID, SeriesInstanceUID, SOPInstanceUID) "
+            "VALUES(?, ?, ?)",
+            instance_data,
         )
+        self._db.commit()
```

Comment:
You may want to consider managing transactions using [context manager](https://docs.python.org/3/library/sqlite3.html#sqlite3-connection-context-manager).

Reply:
The same would apply to several other sections of the code.

Reply:
Nice, I didn't know about this. Implemented in b7f232f